### PR TITLE
[cleanup] for disentangling coqc and the stm

### PR DESCRIPTION
**Pull request by gares**
The final objective is to have a compiler which does not use the STM.

This pr is a simple cleanup which moves common code into files, to be later relocated in a common library (shared between the current coqc and the simpler version). In particular color handling and filename extension checking.

In passing we move the check for pending proofs to a non-legacy API, so that the simple compiler does not need to use it, and we document the APIs we now expose.

### CODE REVIEW

File: toplevel/common_compile.ml [hunk 1]
```diff
@@ -54,4 +54,19 @@ let ensure_exists_with_prefix ~src:f_in ~tgt:f_out ~src_ext:src_suffix ~tgt_ext:
...
     | Some f -> ensure tgt_suffix long_f_dot_src f in
-  long_f_dot_src, long_f_dot_tgt
\ No newline at end of file
+  long_f_dot_src, long_f_dot_tgt
+
+let ensure_no_pending_proofs ~filename s =
+  match s.Vernacstate.lemmas with
+  | Some lemmas ->
+      let pfs = Vernacstate.LemmaStack.get_all_proof_names lemmas in
+      if not (CList.is_empty pfs) then
```
[the remark this answers is not in the document]
Comment (gares) [reply toplevel/common_compile.ml]: ok

**Comment by ejgallego** [comment]
> The final objective is to have a compiler which does not use the STM.

Do we all agree we want that? Shouldn't that then be coordinated with #16190 ?

**Comment by gares** [comment]
I [asked](https://github.com/coq/coq/pull/16190#issuecomment-1173963983) for some reusable code in that PR announcing this work, indeed. I can't base any smaller step like this one on that PR since it has only one large commit which removes too many things.

The bigger plan with @maximedenes  and @SkySkimmer is to have a minimalistic compiler which can be used to test in CI the separation of parsing and execution before that tricky code gets in. Today "just parsing" is doable by hacking the STM (and are are running that in CI for now) but we can't easily tests that "parse;parser;exec;exec" is the same as "parse;exec;parse;exec" since that requires too much STM hacking even for me. With a simpler main compiler loop we can surely test this.

It is unclear to me if this compiler can replace coqc one day. For sure not completely, but maybe we can have a simple compiler and a legacy one which handles -vio and -vos (and links the STM)... to be seen. Do you have thoughts about this?

I hope this PR makes sense anyway, even if we stay as we are. To me having coqc use stuff in coqtop (like colors) makes little sense, as much as having coqc use the legacy API in vernacstate which was added for the STM.

**Comment by SkySkimmer** [comment]
>It is unclear to me if this compiler can replace coqc one day (and if we want it in the repo). For sure It won't completely replace it, but maybe we can have a simple compiler and a legacy one which handles -vio and -vos (and links the STM)... to be seen. Do you have thoughts about this?

I think parsing separation should also make -vos much simpler. Not sure about -vio.

**Comment by ejgallego** [comment]
> I [asked](https://github.com/coq/coq/pull/16190#issuecomment-1173963983) for some reusable code in that PR announcing this work, indeed. I can't base any smaller step like this one on that PR since it has only one large commit which removes too many things.

Ok, I misunderstood what you were asking in that comment. You asked for a commit touching `vernac`, but that PR doesn't touch vernac in a good way. I've removed the deletions for now on that PR, and pushed the cleanups to other branch.

The compiler right now fails in some tests as it doesn't respect the invariant that close_proof imposes, but that's easy to fix if it turns out we want it.

> The bigger plan with @maximedenes and @SkySkimmer is to have a minimalistic compiler which can be used to test in CI the separation of parsing and execution before that tricky code gets in. Today "just parsing" is doable by hacking the STM (and are are running that in CI for now) but we can't easily test that "parse;parser;exec;exec" is the same as "parse;exec;parse;exec" since that requires too much STM hacking even for me. With a simpler main compiler loop we can surely test this.

Where is this plan described? I think we should discuss it, I don't see how not using previous work would be a good strategy here.

I think we should first work on making the parsing state handling explicit on the vernac interpretation as we did for the proofs, then separation comes naturally.

> It is unclear to me if this compiler can replace coqc one day (and if we want it in the repo). For sure It won't completely replace it, but maybe we can have a simple compiler and a legacy one which handles -vio and -vos (and links the STM)... to be seen. Do you have thoughts about this?

Indeed I don't have a clear idea on what to do here; but in order to evaluate this PR we should first answer to the first question? If the simple compiler is not going to live in this repos, not much refactoring may be needed. There is a tiny amount of overlapping code, but that it is.

If you ask me about a thing that would be great to have, I think it'd be to get the meta command out of the `Vernacexpr.vernac_expr` AST, and delegate them to the interp loop. As it is today, document managers and the simple compiler will ideally want to hook at the `Vernacinterp` level.

**Comment by ejgallego** [comment]
> I think parsing separation should also make -vos much simpler. Not sure about -vio.

How?

**Comment by ejgallego** [comment]
> By trimming tactics out and replace Qed by Admitted. This is how people were doing it before it was coded as a vio special case.

I mean, you still need to execute the rest. So how does parsing-only mode help with that? I guess you mean it helps handling possible parsing-side effects inside proofs?

If that's the case, indeed separation is too strong, we just need to classify the commands properly.

(If we just "trim" tactics, then we can do this today easily thanks to the work on typed_vernac)

**Comment by ejgallego** [comment]
> Well, you can separate concerns. You parse, then you patch, then you run.

Yup, that's a way to do it, but I can do it today just fine too, I parse, then I skip, and I run.

It is a long story but actually that's how coq-lsp handles broken proofs and works very well.

File: toplevel/common_compile.mli [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+val ensure_exists : string -> unit
+
+(* [ensure_exists_with_prefix src tgt src_ext tgt_ext] checks that [src] exists
+   (if needed adding the extension) and that [tgt] exists (if needed adding the
+   extension) and returns [src] and [tgt] with their respective extensions.
+   If [tgt] is [None], then it defaults to [src] (with [tgt_ext] as extension).
+   [src_ext] and [tgt_ext] are expected to begin with dot, eg [".v"]. *)
+val ensure_exists_with_prefix : src:string -> tgt:string option -> src_ext:string -> tgt_ext:string -> string * string
+
+(* [chop_extension f] is like Filename.chop_extension but fail safe *)
```
Comment (SkySkimmer) [inline toplevel/common_compile.mli:29]: Should it really have the same name then?

File: toplevel/colors.ml [hunk 3]
```diff
@@ -0,0 +1,79 @@
+(************************************************************************)
```
Comment (SkySkimmer) [inline toplevel/colors.ml:1]: This file seems to be about calling topfmt apis, maybe it should be inlined in topfmt?

**Review by SkySkimmer — APPROVED** [review]

File: toplevel/common_compile.mli (hunk 2 above)
Comment (gares) [reply toplevel/common_compile.mli:29]: done

**Comment by gares** [comment]
all done, thanks @Alizter and @SkySkimmer

**Review by Alizter — APPROVED** [review]

File: toplevel/colors.ml (hunk 3 above)
Comment (gares) [reply toplevel/colors.ml:1]: I was wrong, coqargs is in sysinit/ and it is really about arguments common to most components (it looks weird to put in there something called `parse_extra_colors`.

I'd leave the file where it is in this PR, if you want to chop it, I'm not against, but does not seem to bring much to me.

**Comment by jfehrle** [comment]
> The final objective is to have a compiler which does not use the STM.

Will the compiler will still support async/parallel computation? (I hope so.)

**Comment by gares** [comment]
My plan is not to ditch the current compiler but to clearly distinguish the following cases:
- `coqc` (no STM, called `scoqc` #16190)
- `coqc -vos`, `coqc -vio`, `coqc -async-proos on`, ... (with STM)
- `coqidetop` (with STM)

Then test the branch which separates parsing from execution (before full integration) with the simple `coqc` (too complex to do this with the STM, which is in charge of interleaving parsing and execution).

Finally, build a `vscoqtop` which does not depend on the STM, that takes advantage of the separation of parsing and execution and eventually does parallel/async proofs without Threads but using an event loop based on select.

So, from the user perspective, no feature disappears.

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**